src/Swarm/Util/WindowedCounter.hs

```diff
@@ -0,0 +1,179 @@
+-- of the UI view.
+-- However, we also store the largest element of the window separately from the 'Set' so that
+-- we can compare against it for a @O(1)@ short-circuited path once every member ages out.
+--
+-- The maximum number of elements ever stored in the 'Set' will be the width of the nominal
+-- span, even after some protracted failure to "garbage collect".
+data WindowedCounter a = WindowedCounter
+  { _members :: Set a
+  , _lastLargest :: Maybe a
+  -- ^ NOTE: It is possible that '_lastLargest' may not exist in the 'Set'.
+  , _nominalSpan :: Int
+  -- ^ Data retention window. This value is guaranteed positive by the smart constructor.
+  }
+  deriving (Eq, Show)
+
+-- Automatically deriving 'FromJSON' circumvents the protection offered by "smart constructors",
+-- and the 'ToJSON' instance may expose internal details.
+-- Therefore, we write our own custom implementations.
+instance ToJSON (WindowedCounter a) where
+  toJSON (WindowedCounter membersSet _lastLargest nominalSpan) =
+    object
+      [ "members" .= Set.size membersSet
+      , "span" .= nominalSpan
+      ]
+
+instance FromJSON (WindowedCounter a) where
+  parseJSON = withObject "WindowedCounter" $ \v -> do
+    s <- v .: "span"
+    return $ mkWindow s
+
+-- | NOTE: We take the absolute value of the "window span" argument
+-- so that we can make guarantees about the output of 'getOccupancy'.
+mkWindow ::
+  -- | window span
+  Int ->
+  WindowedCounter a
+mkWindow = WindowedCounter Set.empty Nothing . abs
+
+-- | Return the ratio of {members in the window} to the {integral span
+-- represented by the window}.
+--
+-- The "current time" should be at least as large as the largest
+-- element of the window.
+--
+-- A fully-contiguous collection of ticks would have an occupancy ratio of @1@.
+--
+-- == Unit interval guarantee
+-- The returned ratio is /guaranteed/ to lie on the unit interval, because:
+--
+-- * Both the numerator and denominator of the ratio are guaranteed positive, and
+-- * 'discardGarbage' guarantees that the set size is less than or equal to
+--   the nominal span.
+getOccupancy ::
+  (Ord a, Offsettable a) =>
+  -- | current time
+  a ->
+  WindowedCounter a ->
+  Double
+getOccupancy currentTime wc@(WindowedCounter s lastLargest nominalSpan) =
+  if Set.null s || maybe False (< referenceTick) lastLargest
+    then 0
+    else fromIntegral (Set.size culledSet) / fromIntegral nominalSpan
+ where
+  referenceTick = offsetBy (negate nominalSpan) currentTime
+  -- Cull the window according to the current time
+  WindowedCounter culledSet _ _ = discardGarbage currentTime wc
+
+-- | Invocations of this function shall be guarded externally
+-- by the conditions meant to be tracked in the window.
+--
+-- Proper usage dictates that the value inserted should always
+-- be at least as large as the current largest element of the set.
+-- If not, the insertion has no effect.
```

Comment:
This comment doesn't seem to correspond to what the code actually does.  It seems to me like inserting a value less than the current largest element does actually have an effect, since it gets added to the set (though the `lastLargest` field does not change).  I don't have a strong opinion about which behavior would be better.  I guess I lean towards keeping the current behavior (and changing this comment) since I see no particular reason to disallow "out of order" insertion.

Reply:
Corrected the comment.